src/parser/scanner_support.c

```diff
@@ -145,23 +215,111 @@ set_plan_include_hint (bool is_include)
   plan_include_hint = is_include;
 }
 
-void
-print_hint_dump (FILE * output)
+static void
+print_hint_stmt (FILE * output, int stmt_idx)
 {
   if (s_hint_msg.msg_ptr && *s_hint_msg.msg_ptr)
     {
-      fputs ("\nQuery hints:\n", output);
+      STMT_HINT *hint_ptr;
 
-      s_hint_msg.msg_ptr[s_hint_msg.m_used] = '\0';
-      fprintf (output, "%s", s_hint_msg.msg_ptr);
+      for (hint_ptr = &s_hint_msg.m_hint_head; hint_ptr; hint_ptr = hint_ptr->next)
+	{
+	  if (hint_ptr->stmt_no == stmt_idx)
+	    {
+	      break;
+	    }
+	}
+
+      if (hint_ptr)
+	{
+	  fputs ("\nQuery hints:\n", output);
+
+	  s_hint_msg.msg_ptr[s_hint_msg.m_used] = '\0';
+	  fprintf (output, "%s", s_hint_msg.msg_ptr + hint_ptr->offset);
+	  hint_ptr->stmt_no = -2;	/* change to no more use */
+	}
     }
+}
 
-  s_hint_msg.reset ();
+void
+print_hint_dump (FILE * hint_dump_fp, PARSER_CONTEXT * parser, PT_NODE * select_node)
+{
+  int stmt_idx = 0;
+
+  while (stmt_idx < parser->statement_number)
+    {
+      if (parser->statements[stmt_idx] == NULL)
+	{
+	  stmt_idx++;
+	}
+      else
+	{
+	  if (--s_hint_msg.m_select_cnt == 0)
+	    {
+	      print_hint_stmt (hint_dump_fp, stmt_idx);
+	    }
+	  break;
+	}
+    }
+}
+
+static void
+calc_hint_select_count (PT_NODE * stmt, int &select_cnt)
+{
+  switch (stmt->node_type)
+    {
+    case PT_SELECT:
+      select_cnt++;
+      break;
+    case PT_DIFFERENCE:
+    case PT_INTERSECTION:
+    case PT_UNION:
+      calc_hint_select_count (stmt->info.query.q.union_.arg1, select_cnt);
+      calc_hint_select_count (stmt->info.query.q.union_.arg2, select_cnt);
+      break;
+    default:
+      break;
+    }
+}
+
+void
+calc_hint_statement_count (PT_NODE * statement)
+{
+  s_hint_msg.m_select_cnt = 0;
+  calc_hint_select_count (statement, s_hint_msg.m_select_cnt);
 }
 
 #define HINT_LEAD_CHAR_SIZE (129)
 static u_char hint_table_lead_offset[HINT_LEAD_CHAR_SIZE] = { 0, };
 
+struct s_string
+{
+  int size;
+  char *ptr;
+  char buf[256];
+
+    s_string ()
+  {
+    size = sizeof (buf);
+    ptr = buf;
+    buf[0] = 0x00;
+  }
+
+  void reset ()
+  {
+    if (ptr != buf)
+      {
+	free (ptr);
+      }
+
+    ptr = buf;
+    size = sizeof (buf);
+    buf[0] = 0x00;
+  }
+};
+
+static struct s_string tempory_hint_string;
```

Comment:
How about using `std::string` instead of defining this new struct? `std::string` seems to have the same functionality as the `s_string`